### PR TITLE
escape newlines in request headers

### PR DESCRIPTION
**Pull request by danielhochman**
TODO:
 - [ ] tests for `replaceAll()`

### CODE REVIEW

File: source/common/common/utility.cc [hunk 1]
```diff
@@ -107,3 +107,16 @@ const std::string& StringUtil::valueOrDefault(const std::string& input,
                                               const std::string& default_value) {
   return input.empty() ? default_value : input;
 }
+
+std::string StringUtil::replaceAll(const std::string& input, const std::string& search,
+                                   const std::string& replacement) {
+  std::string input_ = input;
```
Comment (RomanDzhabarov) [inline source/common/common/utility.cc]: minor: x_ is naming convention for class fields. result or something like that would work here.

File: source/common/common/utility.h [hunk 2]
```diff
@@ -98,4 +98,7 @@ class StringUtil {
    */
   static const std::string& valueOrDefault(const std::string& input,
                                            const std::string& default_value);
+
+  static std::string replaceAll(const std::string& input, const std::string& search,
```
Comment (RomanDzhabarov) [inline source/common/common/utility.h]: comment on method

File: source/common/http/access_log/access_log_formatter.h [hunk 3]
```diff
@@ -114,7 +115,7 @@ class RequestHeaderFormatter : public Formatter, HeaderFormatter {
 class ResponseHeaderFormatter : public Formatter, HeaderFormatter {
 public:
   ResponseHeaderFormatter(const std::string& main_header, const std::string& alternative_header,
-                          const Optional<size_t>& max_length);
+                          const Optional<size_t>& max_length, bool);
```
Comment (RomanDzhabarov) [inline source/common/http/access_log/access_log_formatter.h]: if the last param not used, i'd just remove that.

File: source/common/common/utility.cc [hunk 4]
```diff
@@ -107,3 +107,16 @@ const std::string& StringUtil::valueOrDefault(const std::string& input,
...
+
+  size_t pos = 0;
+  while ((pos = input_.find(search, pos)) != std::string::npos) {
```
Comment (mattklein123) [inline source/common/common/utility.cc]: Just use http://en.cppreference.com/w/cpp/algorithm/replace

Comment (danielhochman) [inline source/common/common/utility.cc]: It only works for chars, not strings as far as I can tell.

**Comment by mattklein123** [comment]
I'm actually not sure at this point that we can get a newline inside the header fields. This whole area is kind of complicated:
https://tools.ietf.org/html/draft-ietf-httpbis-p1-messaging-26#section-3.2
https://tools.ietf.org/html/draft-ietf-httpbis-http2-17#section-8.1.2

I would talk to @jpinner-lyft he loves this topic.

If there is any chance of this happening I think I would probably just replace '\n' with ' ' since its more efficient and simpler in general.

**Comment by jpinner-lyft** [comment]
In HTTP/1.x this should be impossible assuming the parser is handling folded header lines correctly (the parser should replace the obs-fold characters along with any multiple tab or space characters with a single space).

In HTTP/2 the header value is considered opaque, but, you should scan the emitted value from the compressor and fail any requests that contain illegal characters:
https://tools.ietf.org/html/draft-ietf-httpbis-http2-17#section-10.3

**Comment by mattklein123** [comment]
HTTP/2 is OK. Just check http_parser:
https://github.com/nghttp2/nghttp2/blob/master/lib/nghttp2_helper.c#L495

**Comment by danielhochman** [comment]
won't fix. parser properly handles newlines in headers and converts them to spaces.

**Pull request closed without merging**